src/main/kotlin/cacophony/grammars/Analysis.kt

```diff
@@ -1,44 +1,169 @@
 package cacophony.grammars
 
 import cacophony.automata.DFA
+import kotlin.math.min
 
-typealias DFAStateReference<StateType, AtomType, ResultType> = Pair<StateType, DFA<StateType, AtomType, ResultType>>
+typealias DFAStateReference<StateType, SymbolType, ResultType> =
+    Pair<StateType, DFA<StateType, SymbolType, ResultType>>
 
-fun <StateType, SymbolType> findNullable(
-    automata: Map<SymbolType, DFA<StateType, SymbolType, Production<SymbolType>>>,
-): Collection<DFAStateReference<StateType, SymbolType, Production<SymbolType>>> {
-    TODO("Not implemented")
+sealed class Entity<out Q, out A, out R> {
+    data class Ref<Q, A, R>(val ref: DFAStateReference<Q, A, R>) : Entity<Q, A, R>()
+
+    data class Sym<A>(val sym: A) : Entity<Nothing, A, Nothing>()
 }
 
-fun <StateType, SymbolType> findFirst(
-    automata: Map<SymbolType, DFA<StateType, SymbolType, Production<SymbolType>>>,
-    nullable: Collection<DFAStateReference<StateType, SymbolType, Production<SymbolType>>>,
-): Map<DFAStateReference<StateType, SymbolType, Production<SymbolType>>, Collection<SymbolType>> {
+typealias StateToSymbolsMap<StateType, SymbolType, ResultType> =
+    Map<DFAStateReference<StateType, SymbolType, ResultType>, Collection<SymbolType>>
+
+fun <StateType, SymbolType, ResultType> findNullable(
+    automata: Map<SymbolType, DFA<StateType, SymbolType, ResultType>>,
+): Collection<DFAStateReference<StateType, SymbolType, ResultType>> {
     TODO("Not implemented")
 }
 
-fun <StateType, SymbolType> findFollow(
-    automata: Map<SymbolType, DFA<StateType, SymbolType, Production<SymbolType>>>,
-    nullable: Collection<DFAStateReference<StateType, SymbolType, Production<SymbolType>>>,
-    first: Map<DFAStateReference<StateType, SymbolType, Production<SymbolType>>, Collection<SymbolType>>,
-): Map<DFAStateReference<StateType, SymbolType, Production<SymbolType>>, Collection<SymbolType>> {
+class FindFirstImpl<StateType, SymbolType, ResultType>(
```

Comment:
what's the advantage of creating this class instead of just putting the code in `findFirst()` function like it was declared before?

Reply:
I had some problems to make it compile otherwise. Fixed now.